userdata.tpl

```diff
@@ -2,9 +2,134 @@ MIME-Version: 1.0
 Content-Type: multipart/mixed; boundary="//"
 
 --//
-Content-Type: text/x-shellscript; charset="us-ascii"
-#!/bin/bash
+Content-Type: text/x-shellscript
+Content-Type: charset="us-ascii"
 
 ${before_cluster_joining_userdata}
 
---//--
+cat << 'EOF' > /etc/eks/extra_args.sh
+#!/usr/bin/env bash
+set -ex +u
+
+function prepare_lines() {
+  local KUBELET_EXTRA_ARGS="$1"
+
+  # Break each argument into its own line, and replace the first '=' with '%' for easier processing later
+  # Example output:
+  # --register-with-taints%dedicated=test:NoSchedule
+  # --node-labels%eks.amazonaws.com/nodegroup=test-mng,other.label=other
+  #
+  local EXTRA_ARGS_LINES=$(echo "$KUBELET_EXTRA_ARGS" | xargs -d' ' -I{} bash -c 'echo "{}" | sed "0,/=/s//%/"')
+
+  # Remove blank lines
+  # Example output:
+  # --register-with-taints%dedicated=test:NoSchedule
+  # --node-labels%eks.amazonaws.com/nodegroup=test-mng,other.label=other
+  local EXTRA_ARGS_LINES=$(echo "$EXTRA_ARGS_LINES" | sed '/^$/d')
+
+  echo "$EXTRA_ARGS_LINES"
+}
+
+# 30-kubelet-extra-args.conf example:
+# [Service]
+# Environment='KUBELET_EXTRA_ARGS=--register-with-taints="dedicated=test:NoSchedule" --node-labels=eks.amazonaws.com/nodegroup=test-mng,other.label=other'
+
+# Get the last line and cut out the part between the single quotes
+# Example output:
+# KUBELET_EXTRA_ARGS=--register-with-taints="dedicated=test:NoSchedule" --node-labels=eks.amazonaws.com/nodegroup=test-mng,other.label=other
+ENVIRONMENT=$(tail -n1 /etc/systemd/system/kubelet.service.d/30-kubelet-extra-args.conf | cut -d\' -f2)
```

Comment:
This is brittle, in that it requires that the last line of the file be desired line and will break if/when the default changes.